File: events/2015/21/solution.go
package main

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
	"time"
	"yod_aoc/utils/go/aoc_utils"
)

type PlayerStats struct {
	Health int
	Armor  int
	Damage int
	Spent  int
}

type ShopItem struct {
	Cost     int
	Damage   int
	Armor    int
	Category ShopCategory
}

type ShopCategory string

const (
	CategoryWeapon ShopCategory = "Weapon"
	CategoryArmor  ShopCategory = "Armor"
	CategoryRing   ShopCategory = "Ring"
)

func getShopItems() map[string]ShopItem {
	items := make(map[string]ShopItem)

	items["Dagger"] = ShopItem{Cost: 8, Damage: 4, Armor: 0, Category: CategoryWeapon}
	items["Shortsword"] = ShopItem{Cost: 10, Damage: 5, Armor: 0, Category: CategoryWeapon}
	items["Warhammer"] = ShopItem{Cost: 25, Damage: 6, Armor: 0, Category: CategoryWeapon}
	items["Longsword"] = ShopItem{Cost: 40, Damage: 7, Armor: 0, Category: CategoryWeapon}
	items["Greataxe"] = ShopItem{Cost: 74, Damage: 8, Armor: 0, Category: CategoryWeapon}

	// Armor
	items["Leather"] = ShopItem{Cost: 13, Damage: 0, Armor: 1, Category: CategoryArmor}
	items["Chainmail"] = ShopItem{Cost: 31, Damage: 0, Armor: 2, Category: CategoryArmor}
	items["Splitmail"] = ShopItem{Cost: 53, Damage: 0, Armor: 3, Category: CategoryArmor}
	items["Bandemail"] = ShopItem{Cost: 75, Damage: 0, Armor: 4, Category: CategoryArmor}
	items["Platemail"] = ShopItem{Cost: 102, Damage: 0, Armor: 5, Category: CategoryArmor}

	// Rings
	items["Damage +1"] = ShopItem{Cost: 25, Damage: 1, Armor: 0, Category: CategoryRing}
	items["Damage +2"] = ShopItem{Cost: 50, Damage: 2, Armor: 0, Category: CategoryRing}
	items["Damage +3"] = ShopItem{Cost: 100, Damage: 3, Armor: 0, Category: CategoryRing}
	items["Defense +1"] = ShopItem{Cost: 20, Damage: 0, Armor: 1, Category: CategoryRing}
	items["Defense +2"] = ShopItem{Cost: 40, Damage: 0, Armor: 2, Category: CategoryRing}
	items["Defense +3"] = ShopItem{Cost: 80, Damage: 0, Armor: 3, Category: CategoryRing}

	return items
}

func runBattle(player PlayerStats, enemy PlayerStats) bool {
	turn := 0
	playerDead := false
	enemyDead := false

	for !playerDead && !enemyDead {
		turn += 1

		if turn%2 == 0 {
			dealtDamage := math.Max(float64(enemy.Damage-player.Armor), float64(1))
			player.Health -= int(dealtDamage)

			if player.Health <= 0 {
				playerDead = true
			}
		} else {
			dealtDamage := math.Max(float64(player.Damage-enemy.Armor), float64(1))
			enemy.Health -= int(dealtDamage)

			if enemy.Health <= 0 {
				enemyDead = true
			}
		}
	}

	return playerDead
}

func getPossibleLoadouts(shopItems map[string]ShopItem, startingHealth int) []PlayerStats {
	possibleLoadouts := []PlayerStats{}

	weapons := make(map[string]ShopItem)
	armors := make(map[string]ShopItem)
	rings := make(map[string]ShopItem)

	for key, value := range shopItems {
		if value.Category == CategoryWeapon {
			weapons[key] = value
		} else if value.Category == CategoryArmor {
			armors[key] = value
		} else if value.Category == CategoryRing {
			rings[key] = value
		}
	}

	weaponLoadouts := []PlayerStats{}
	for _, weapon := range weapons {
		loadout := PlayerStats{
			Health: startingHealth,
			Damage: weapon.Damage,
			Armor:  0,
			Spent:  weapon.Cost,
		}
		weaponLoadouts = append(weaponLoadouts, loadout)
		possibleLoadouts = append(possibleLoadouts, loadout)
	}

	armorLoadouts := append([]PlayerStats{}, weaponLoadouts...)
	for _, armor := range armors {
		for _, weapon := range weaponLoadouts {
			loadout := PlayerStats{
				Health: startingHealth,
				Damage: weapon.Damage,
				Armor:  armor.Armor,
				Spent:  weapon.Spent + armor.Cost,
			}
			armorLoadouts = append(armorLoadouts, loadout)
			possibleLoadouts = append(possibleLoadouts, loadout)
		}
	}

	for _, ring := range rings {
		for _, armor := range armorLoadouts {
			possibleLoadouts = append(possibleLoadouts, PlayerStats{
				Health: startingHealth,
				Damage: armor.Damage + ring.Damage,
				Armor:  armor.Armor + ring.Armor,
				Spent:  armor.Spent + ring.Cost,
			})
		}
	}

	for name1, ring1 := range rings {
		for name2, ring2 := range rings {
			if name1 == name2 {
				continue
			}

			for _, armor := range armorLoadouts {
				possibleLoadouts = append(possibleLoadouts, PlayerStats{
					Health: startingHealth,
					Damage: armor.Damage + ring1.Damage + ring2.Damage,
					Armor:  armor.Armor + ring1.Armor + ring2.Armor,
					Spent:  armor.Spent + ring1.Cost + ring2.Cost,
				})
			}
		}
	}

	return possibleLoadouts
}

// ----------------------------------------------------------------------------------------------------
// | Part 1
// ----------------------------------------------------------------------------------------------------
func part1(input []string) int {
	Enemy := PlayerStats{Health: 0, Damage: 0, Armor: 0, Spent: 0}

	for _, line := range input {
		if len(line) == 0 {
			continue
		}

		if strings.Contains(line, "Hit Points") {
			health, err := strconv.Atoi(strings.Split(line, ": ")[1])
			if err != nil {
				panic(err)
			}
			Enemy.Health = health
		} else if strings.Contains(line, "Damage") {
			damage, err := strconv.Atoi(strings.Split(line, ": ")[1])
			if err != nil {
				panic(err)
			}
			Enemy.Damage = damage
		} else if strings.Contains(line, "Armor") {
			armor, err := strconv.Atoi(strings.Split(line, ": ")[1])
			if err != nil {
				panic(err)
			}
			Enemy.Armor = armor
		}
	}

	loadouts := getPossibleLoadouts(getShopItems(), 100)

	lowestSpend := 0
	initialSet := false
	for _, loadout := range loadouts {
		battleResult := runBattle(loadout, PlayerStats{Health: Enemy.Health, Damage: Enemy.Damage, Armor: Enemy.Armor, Spent: 0})

		if !battleResult {
			if lowestSpend > loadout.Spent || !initialSet {
				initialSet = true
				lowestSpend = loadout.Spent
			}
		}
	}

	return lowestSpend
}

// ----------------------------------------------------------------------------------------------------
// | Part 2
// ----------------------------------------------------------------------------------------------------
func part2(input []string) int {
	Enemy := PlayerStats{Health: 0, Damage: 0, Armor: 0, Spent: 0}

	for _, line := range input {
		if len(line) == 0 {
			continue
		}

		if strings.Contains(line, "Hit Points") {
			health, err := strconv.Atoi(strings.Split(line, ": ")[1])
			if err != nil {
				panic(err)
			}
			Enemy.Health = health
		} else if strings.Contains(line, "Damage") {
			damage, err := strconv.Atoi(strings.Split(line, ": ")[1])
			if err != nil {
				panic(err)
			}
			Enemy.Damage = damage
		} else if strings.Contains(line, "Armor") {
			armor, err := strconv.Atoi(strings.Split(line, ": ")[1])
			if err != nil {
				panic(err)
			}
			Enemy.Armor = armor
		}
	}

	loadouts := getPossibleLoadouts(getShopItems(), 100)

	highestSpend := 0
	initialSet := false
	for _, loadout := range loadouts {
		battleResult := runBattle(loadout, PlayerStats{Health: Enemy.Health, Damage: Enemy.Damage, Armor: Enemy.Armor, Spent: 0})

		if battleResult {
			if highestSpend < loadout.Spent || !initialSet {
				initialSet = true
				highestSpend = loadout.Spent
			}
		}
	}

	return highestSpend
}

// ----------------------------------------------------------------------------------------------------
// | Main Function
// ----------------------------------------------------------------------------------------------------
func main() {
	// Read the arguments
	args := os.Args[1:]
	var fileName string
	if len(args) > 0 {
		fileName = args[0]
	} else {
		fileName = "input.txt"
	}
	// Start the timer!
	startTime := time.Now()

	// Read the input to an array of strings
	path := aoc_utils.GetFilePath(2015, 21, fileName)
	content, err := aoc_utils.ReadInputAsArray(path)
	if err != nil {
		fmt.Println("Error reading input:", err)
		return
	}

	// Run the Parts
	p1Result := aoc_utils.MeasurePerformance(func() interface{} {
		return part1(content)
	})

	p2Result := aoc_utils.MeasurePerformance(func() interface{} {
		return part2(content)
	})

	// End the Timer!
	endTime := time.Now()
	duration := endTime.Sub(startTime)

	resultsForDay := aoc_utils.AOCDayResults{
		Year:     2015,
		Day:      21,
		Part1:    p1Result,
		Part2:    p2Result,
		Duration: duration.Nanoseconds(),
		Timestamp: aoc_utils.AOCTimestamp{
			Start: startTime.Format(time.RFC3339),
			End:   endTime.Format(time.RFC3339),
		},
	}
	jsonResults, _ := json.Marshal(resultsForDay)
	fmt.Println(string(jsonResults))
}

File: events/2015/21/solution.rs
use aoc_utils::read_input_as_array;
use aoc_utils::measure_performance;
use aoc_utils::AOCDayResults;
use aoc_utils::AOCTimestamp;
use chrono::{DateTime, Utc};
use std::collections::HashMap;
use std::cmp::max;

#[derive(Debug, Clone)]
struct PlayerStats {
  health: i32,
  armor: i32,
  damage: i32,
  spent: i32
}

#[derive(Debug, PartialEq, Clone)]
enum ShopCategory {
  Weapon,
  Armor,
  Ring
}

#[derive(Debug, Clone)]
struct ShopItem {
  cost: i32,
  damage: i32,
  armor: i32,
  category: ShopCategory
}

fn get_shop_items() -> HashMap<String, ShopItem>{
  let mut items: HashMap<String, ShopItem> = HashMap::new();

  // Weapons
  items.insert("Dagger".to_string(), ShopItem {cost: 8, damage: 4, armor: 0, category: ShopCategory::Weapon});
  items.insert("Shortsword".to_string(), ShopItem {cost: 10, damage: 5, armor: 0, category: ShopCategory::Weapon});
  items.insert("Warhammer".to_string(), ShopItem {cost: 25, damage: 6, armor: 0, category: ShopCategory::Weapon});
  items.insert("Longsword".to_string(), ShopItem {cost: 40, damage: 7, armor: 0, category: ShopCategory::Weapon});
  items.insert("Greataxe".to_string(), ShopItem {cost: 74, damage: 8, armor: 0, category: ShopCategory::Weapon});

  // Armor
  items.insert("Leather".to_string(), ShopItem {cost: 13, damage: 0, armor: 1, category: ShopCategory::Armor});
  items.insert("Chainmail".to_string(), ShopItem {cost: 31, damage: 0, armor: 2, category: ShopCategory::Armor});
  items.insert("Splitmail".to_string(), ShopItem {cost: 53, damage: 0, armor: 3, category: ShopCategory::Armor});
  items.insert("Bandemail".to_string(), ShopItem {cost: 75, damage: 0, armor: 4, category: ShopCategory::Armor});
  items.insert("Platemail".to_string(), ShopItem {cost: 102, damage: 0, armor: 5, category: ShopCategory::Armor});

  // Rings
  items.insert("Damage +1".to_string(), ShopItem {cost: 25, damage: 1, armor: 0, category: ShopCategory::Ring});
  items.insert("Damage +2".to_string(), ShopItem {cost: 50, damage: 2, armor: 0, category: ShopCategory::Ring});
  items.insert("Damage +3".to_string(), ShopItem {cost: 100, damage: 3, armor: 0, category: ShopCategory::Ring});
  items.insert("Defense +1".to_string(), ShopItem {cost: 20, damage: 0, armor: 1, category: ShopCategory::Ring});
  items.insert("Defense +2".to_string(), ShopItem {cost: 40, damage: 0, armor: 2, category: ShopCategory::Ring});
  items.insert("Defense +3".to_string(), ShopItem {cost: 80, damage: 0, armor: 3, category: ShopCategory::Ring});

  return items;
}

fn run_battle(mut player: PlayerStats, mut enemy: PlayerStats) -> bool {
  let mut turn: i32 = 0;
  let mut player_dead: bool = false;
  let mut enemy_dead: bool = false;

  while !player_dead && !enemy_dead {
    turn += 1;

    if turn % 2 == 0 {
      let dealt_damage = max(enemy.damage - player.armor, 1);
      player.health -= dealt_damage;

      if player.health <= 0 {
        player_dead = true;
      }
    } else {
      let dealt_damage = max(player.damage - enemy.armor, 1);
      enemy.health -= dealt_damage;

      if enemy.health <= 0 {
        enemy_dead = true;
      }
    }
  }

  return player_dead
}

fn get_possible_loadouts(shop_items: &HashMap<String, ShopItem>, starting_health: i32) -> Vec<PlayerStats> {
  let mut possible_loadouts: Vec<PlayerStats> = Vec::new();

  let weapons: HashMap<String, ShopItem> = shop_items
    .iter()
    .filter(|(_, value)| value.category == ShopCategory::Weapon)
    .map(|(key, value)| (key.clone(), value.clone()))
    .collect();

  let armors: HashMap<String, ShopItem> = shop_items
    .iter()
    .filter(|(_, value)| value.category == ShopCategory::Armor)
    .map(|(key, value)| (key.clone(), value.clone()))
    .collect();

  let rings: HashMap<String, ShopItem> = shop_items
    .iter()
    .filter(|(_, value)| value.category == ShopCategory::Ring)
    .map(|(key, value)| (key.clone(), value.clone()))
    .collect();

  let mut weapon_loadouts: Vec<PlayerStats> = Vec::new();
  for (_, weapon) in &weapons {
    let loadout = PlayerStats {
      health: starting_health,
      damage: weapon.damage,
      armor: 0,
      spent: weapon.cost,
    };
    weapon_loadouts.push(loadout.clone());
    possible_loadouts.push(loadout);
  }

  let mut armor_loadouts = weapon_loadouts.clone();
  for (_, armor) in &armors {
    for wl in &weapon_loadouts {
      let loadout = PlayerStats {
        health: starting_health,
        damage: wl.damage,
        armor: armor.armor,
        spent: wl.spent + armor.cost,
      };
      armor_loadouts.push(loadout.clone());
      possible_loadouts.push(loadout);
    }
  }

  for (_, ring) in &rings {
    for ar in &armor_loadouts {
      possible_loadouts.push(PlayerStats {
        health: starting_health,
        damage: ar.damage + ring.damage,
        armor: ar.armor + ring.armor,
        spent: ar.spent + ring.cost,
      });
    }
  }

  for (item1, ring1) in &rings {
    for (item2, ring2) in &rings {
      if item1 == item2 {
        continue;
      }

      for ar in &armor_loadouts {
        possible_loadouts.push(PlayerStats {
          health: starting_health,
          damage: ar.damage + ring1.damage + ring2.damage,
          armor: ar.armor + ring1.armor + ring2.armor,
          spent: ar.spent + ring1.cost + ring2.cost,
        });
      }
    }
  }

  possible_loadouts
}

// ----------------------------------------------------------------------------------------------------
// | Part 1
// ----------------------------------------------------------------------------------------------------
fn part1(input: &Vec<String>) -> i32 {
  let mut enemy: PlayerStats = PlayerStats {health: 0, armor: 0, damage: 0, spent: 0};

  for (_i, line) in input.iter().enumerate() {
    if line.is_empty() {
      continue
    }

    if line.contains("Hit Points") {
      if let Some(value) = line.split(": ").nth(1) {
        enemy.health = value.parse().unwrap_or(0);
      }
    } else if line.contains("Damage") {
      if let Some(value) = line.split(": ").nth(1) {
        enemy.damage = value.parse().unwrap_or(0);
      }
    } else if line.contains("Armor") {
      if let Some(value) = line.split(": ").nth(1) {
        enemy.armor = value.parse().unwrap_or(0);
      }
    }
  }

  let loadouts = get_possible_loadouts(&get_shop_items(), 100);

  let mut lowest_spend = i32::MAX;
  for loadout in &loadouts {
    let enemy_stats = enemy.clone();
    let battle_result = run_battle(loadout.clone(), enemy_stats);

    if !battle_result {
      if lowest_spend > loadout.spent {
        lowest_spend = loadout.spent;
      }
    }
  }
  
  lowest_spend
}

// ----------------------------------------------------------------------------------------------------
// | Part 2
// ----------------------------------------------------------------------------------------------------
fn part2(input: &Vec<String>) -> i32 {
  let mut enemy: PlayerStats = PlayerStats {health: 0, armor: 0, damage: 0, spent: 0};

  for (_i, line) in input.iter().enumerate() {
    if line.is_empty() {
      continue
    }

    if line.contains("Hit Points") {
      if let Some(value) = line.split(": ").nth(1) {
        enemy.health = value.parse().unwrap_or(0);
      }
    } else if line.contains("Damage") {
      if let Some(value) = line.split(": ").nth(1) {
        enemy.damage = value.parse().unwrap_or(0);
      }
    } else if line.contains("Armor") {
      if let Some(value) = line.split(": ").nth(1) {
        enemy.armor = value.parse().unwrap_or(0);
      }
    }
  }

  let loadouts = get_possible_loadouts(&get_shop_items(), 100);

  let mut highest_spend: i32 = 0;
  for loadout in &loadouts {
    let enemy_stats = enemy.clone();
    let battle_result = run_battle(loadout.clone(), enemy_stats);

    if battle_result {
      if highest_spend < loadout.spent {
        highest_spend = loadout.spent;
      }
    }
  }
  
  highest_spend
}

// ----------------------------------------------------------------------------------------------------
// | Main Function
// ----------------------------------------------------------------------------------------------------
fn main() {
  // Read the arguments
  let args = std::env::args().collect::<Vec<String>>();
  let default_file_name = String::from("input.txt");
  let file_name = args.get(1).unwrap_or(&default_file_name);

  // Start the timer
  let start_timestamp = std::time::SystemTime::now();
  let start_time = std::time::Instant::now();
  
  // Read the input
  let input = read_input_as_array(file_name).unwrap();

  // Run the Parts
  let p1_result = measure_performance(part1, &input);
  let p2_result = measure_performance(part2, &input);

  // Stop the timer
  let end_time = std::time::Instant::now();
  let end_timestamp = std::time::SystemTime::now();
  let duration = end_time.duration_since(start_time).as_nanos();

  let results = AOCDayResults {
    year: 2015,
    day: 21,
    part1: p1_result,
    part2: p2_result,
    duration: duration,
    timestamp: AOCTimestamp {
      start: DateTime::<Utc>::from(start_timestamp).to_rfc3339(),
      end: DateTime::<Utc>::from(end_timestamp).to_rfc3339(),
    },
  };

  let json_results = serde_json::to_string(&results).unwrap();
  println!("{}", json_results);
}